### DISCUSSION TITLE
Where should authorization logic go?

### DISCUSSION THREAD

**Discussion by jerrac** [symfony/symfony#46897] [Q&A]
I started a little side project a couple days ago to learn Symfony 6.1. I'm at the point now where I want to control who can see what content. The firewall makes it simple to control access to routes, and that makes sense for some content. But where is the best place to control access to entities?

For example, if I have a list of entities, I'd want admins to see all items, but users should only see the ones they own.

I could see putting the authorization logic in the entity controller, but it could also go into the entity repository. Or even into the template.

Is there a "best practice" around this?

What other kinds of authorization should I think about?

Thanks in advance!

**Comment by gizmola** [comment]
You want to look into writing custom a custom voter:  See https://symfony.com/doc/current/security/voters.html  

I would highly recommend Symfonycasts Security section.  This is discussed in detail here:  https://symfonycasts.com/screencast/symfony-security/custom-voter

This is for Symfony 5, as their Symfony 6 course is still being worked on, but most of the Symfony 5 material is still relevant to 6.  They do have a completed "[Upgrade 5 -> 6](https://symfonycasts.com/screencast/symfony6-upgrade)" course that covers important changes including changes to the security and authentication core.

**Reply by jerrac** [reply]
Thanks. I had read about custom voters, but hadn't seen that specific doc, or connected the dots on how they'd help. That has helped me get a bit farther along.

That said, I'm still making calls to `is_granted` and `denyAccessUnlessGranted` in my controllers, and `is_granted` in my templates. And I've considered putting checks into my entity repository findAll methods as well. So I'm still curious about an overall strategy. Should I try to keep all access/authorization log in my controllers if possible? Should I avoid adding it to repositories? That kind of thing. :)
